File: Agents/java/LightingAgent/src/main/java/com/partyplanning/lightingagent/controllers/RunController.java
package com.partyplanning.lightingagent.controllers;

import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import com.partyplanning.lightingagent.models.AssistantThreadBase;
import com.partyplanning.lightingagent.models.AssistantThreadRun;
import com.partyplanning.lightingagent.services.RunService;
import com.partyplanning.lightingagent.utils.AssistantEventStreamService;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@RestController
@RequestMapping("/api/threads/{threadId}/runs")
public class RunController {

    @Autowired
    private MongoTemplate mongoTemplate;

    @Autowired
    private RunService runService;

    @Autowired
    private AssistantEventStreamService assistantEventStreamService;

    @PostMapping
    public SseEmitter createRun(@PathVariable String threadId) {
        if (threadId == null || threadId.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Thread ID is required.");
        }

        AssistantThreadBase thread = mongoTemplate.findById(threadId, AssistantThreadBase.class, "Threads");
        if (thread == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Thread with ID '" + threadId + "' not found.");
        }

        AssistantThreadRun newRun = new AssistantThreadRun();
        newRun.setThreadId(threadId);
        newRun.setAssistantId("lighting-agent");

        SseEmitter emitter = new SseEmitter(Long.MAX_VALUE);
        ExecutorService sseMvcExecutor = Executors.newSingleThreadExecutor();
        sseMvcExecutor.execute(() -> {
            try {
                assistantEventStreamService.sendEvent(emitter, "thread.run.created", newRun);
                assistantEventStreamService.sendEvent(emitter, "thread.run.queued", newRun);
                assistantEventStreamService.sendEvent(emitter, "thread.run.in_progress", newRun);
                assistantEventStreamService.sendEvent(emitter, "thread.run.step.created", newRun);
                assistantEventStreamService.sendEvent(emitter, "thread.run.step.in_progress", newRun);

                runService.executeRun(emitter, newRun);

                assistantEventStreamService.sendEvent(emitter, "thread.run.completed", newRun);
                assistantEventStreamService.sendEvent(emitter, "thread.run.step.completed", newRun);
                assistantEventStreamService.sendDoneEvent(emitter);

                emitter.complete();
            } catch (Exception e) {
                emitter.completeWithError(e);
            }
        });

        return emitter;
    }
}

File: Agents/dotnet/Shared/Controllers/RunController.cs
using Microsoft.AspNetCore.Mvc;
using MongoDB.Driver;
using PartyPlanning.Agents.Shared.Models;
using Microsoft.Extensions.Options;
using PartyPlanning.Agents.Shared.Config;
using PartyPlanning.Agents.Shared.Services;
using Microsoft.AspNetCore.Http;

namespace PartyPlanning.Agents.Shared.Controllers
{
    /// <summary>
    /// Controller for managing runs within a specific thread.
    /// </summary>
    /// <remarks>
    /// Initializes a new instance of the <see cref="RunController"/> class.
    /// </remarks>
    /// <param name="database">The MongoDB database.</param>
    /// <param name="runService">The run service.</param>
    /// <param name="AgentConfiguration">Discovery information for the agent</param>
    /// <param name="assistantEventStreamService">Provides utilities to manage Assistant API stream events</param>
    [ApiController]
    [Route("/api/threads/{threadId}/runs")]
    public class RunController(
        IMongoDatabase database,
        RunService runService,
        IOptions<AgentConfiguration> AgentConfiguration,
        AssistantEventStreamService assistantEventStreamService
    ) : ControllerBase
    {
        private readonly IMongoCollection<AssistantThreadBase> _threadsCollection = database.GetCollection<AssistantThreadBase>("Threads");
        private readonly RunService _runService = runService;

        /// <summary>
        /// Creates a new run within a specific thread.
        /// </summary>
        /// <param name="threadId">The ID of the thread to create the run in</param>
        /// <returns>Server-sent events stream of the run</returns>
        [HttpPost]
        public async Task<IActionResult> CreateRun(string threadId)
        {
            if (string.IsNullOrEmpty(threadId))
            {
                return BadRequest("Thread ID is required.");
            }


            var thread = await _threadsCollection.Find(t => t.Id == threadId).FirstOrDefaultAsync();

            if (thread == null)
            {
                return NotFound($"Thread with ID '{threadId}' not found.");
            }

            var newRun = new AssistantThreadRun()
            {
                ThreadId = threadId,
                AssistantId = AgentConfiguration.Value.Name,
                CreatedAt = DateTime.UtcNow
            };

            async IAsyncEnumerable<string> RunEventStream()
            {
                yield return assistantEventStreamService.CreateEvent("thread.run.created", newRun);
                yield return assistantEventStreamService.CreateEvent("thread.run.queued", newRun);
                yield return assistantEventStreamService.CreateEvent("thread.run.in_progress", newRun);
                yield return assistantEventStreamService.CreateEvent("thread.run.step.created", newRun);
                yield return assistantEventStreamService.CreateEvent("thread.run.step.in_progress", newRun);

                await foreach (var events in _runService.ExecuteRunAsync(newRun))
                {
                    yield return events;
                }
                
                yield return assistantEventStreamService.CreateEvent("thread.run.completed", newRun);
                yield return assistantEventStreamService.CreateEvent("thread.run.step.completed", newRun);
                yield return assistantEventStreamService.CreateDoneEvent();
            }

            Response.ContentType = "text/event-stream";
            await foreach (var message in RunEventStream())
            {
                await Response.WriteAsync(message);
                await Response.Body.FlushAsync();
            }

            return new EmptyResult();
        }
    }
}
